### ISSUE TITLE
clarify how clusterctl overrides works

### ISSUE THREAD

**Issue by namnx228** [kubernetes-sigs/cluster-api#5818]
**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]
`clusterctl upgrade plan | apply` ignores the overrides repository and fetch the metadata.yaml file from upstream.
Here is my overrides structure:
```sh
.cluster-api/
├── clusterctl.yaml
├── overrides
│   └── infrastructure-metal3
│       └── v1.0.0
└── version.yaml

```
clusterctl.yaml:
```yaml
KUBERNETES_VERSION: v1.22.2
NAMESPACE: metal3
NUM_OF_MASTER_REPLICAS: 3
NUM_OF_WORKER_REPLICAS: 1
NODE_DRAIN_TIMEOUT: 0s
MAX_SURGE_VALUE: 0

## Cluster network
CLUSTER_APIENDPOINT_HOST: 192.168.111.249
CLUSTER_APIENDPOINT_PORT: 6443
POD_CIDR: 192.168.0.0/18
SERVICE_CIDR: 10.96.0.0/12
PROVISIONING_POOL_RANGE_START: 172.22.0.100
PROVISIONING_POOL_RANGE_END: 172.22.0.200
PROVISIONING_CIDR: 24
BAREMETALV4_POOL_RANGE_START: 192.168.111.100
BAREMETALV4_POOL_RANGE_END: 192.168.111.200
BAREMETALV6_POOL_RANGE_START: fd55::100
BAREMETALV6_POOL_RANGE_END: fd55::200
EXTERNAL_SUBNET_V4_PREFIX: 24
EXTERNAL_SUBNET_V4_HOST: 192.168.111.1
EXTERNAL_SUBNET_V6_PREFIX: 64
EXTERNAL_SUBNET_V6_HOST: fd55::1
```
and this is the debug message from `clusterctl upgrade plan`:
```sh
$ clusterctl upgrade plan -v 9
Using configuration File="/home/ubuntu/.cluster-api/clusterctl.yaml"
Checking cert-manager version...
Cert-Manager is already up to date

Checking new release availability...
Fetching File="metadata.yaml" Provider="cluster-api" Type="CoreProvider" Version="v1.0.2"
Fetching File="metadata.yaml" Provider="kubeadm" Type="BootstrapProvider" Version="v1.0.2"
Fetching File="metadata.yaml" Provider="kubeadm" Type="ControlPlaneProvider" Version="v1.0.2"
Fetching File="metadata.yaml" Provider="cluster-api" Type="CoreProvider" Version="v1.0.2"
Fetching File="metadata.yaml" Provider="metal3" Type="InfrastructureProvider" Version="v0.5.3"
```
If the metal3 infrastructure provider metadata.yaml is used from the overrides directory, the last message should be:
`Using Override="metadata.yaml" Provider="infrastructure-metal3" Version="v1.0.0"`

**What did you expect to happen:**
When setting up the overrides directory for a provider, it is expected that clusterctl upgrade plan | apply would use the metadata.yaml file in this overrides directory instead of fetching it from upstream. This behavior is true with clusterctl v0.4.4. However, it is not true with >=v1.0.0.

**Environment:**

- Cluster-api version: >=v1.0.0
- Minikube/KIND version: kind version 0.11.1
- Kubernetes version: (use `kubectl version`):
Client: v1.22.4
Server: v1.22.2
- OS (e.g. from `/etc/os-release`): Ubuntu 20.04.3 LTS
/kind bug 
/area upgrade clusterctl
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by sbueringer** [comment]
I think the clusterctl.yaml has to contain something like this:
```yaml
providers:
- name: "cluster-api"
  type: "CoreProvider"
  url: "/Users/buringerst/.cluster-api/dev-repository/cluster-api/v1.1.99/core-components.yaml"
- name: "kubeadm"
  type: "BootstrapProvider"
  url: "/Users/buringerst/.cluster-api/dev-repository/bootstrap-kubeadm/v1.1.99/bootstrap-components.yaml"
- name: "kubeadm"
  type: "ControlPlaneProvider"
  url: "/Users/buringerst/.cluster-api/dev-repository/control-plane-kubeadm/v1.1.99/control-plane-components.yaml"
- name: "docker"
  type: "InfrastructureProvider"
  url: "/Users/buringerst/.cluster-api/dev-repository/infrastructure-docker/v1.1.99/infrastructure-components.yaml"
overridesFolder: "/Users/buringerst/.cluster-api/dev-repository/overrides"
```

**Comment by namnx228** [comment]
@sbueringer I expect it works even when we don't have the provider field in the `clusterctl.yaml`. With clusterctl <=v1.0.0, it uses the metadata.yaml from overrides repository without the provider field. 
One thing I should add is that `clusterctl init` in my case still use the overrides repository, so only `clusterctl upgrade` has this issue.

**Comment by sbueringer** [comment]
Ah okay, I wasn't aware of that. Let's see what other folks which know more about clusterctl say

**Comment by namnx228** [comment]
cc @fabriziopandini

**Comment by fabriziopandini** [comment]
@ykakarap PTAL

**Comment by ykakarap** [comment]
@namnx228 I see that you mentioned that `clsuterctl init` picked the correct overrides. Can you share more details about this case?
Can you share the version of clusterctl used when running the init command? alos, if you have it can you share the init command you used? did you use specify a version of the provider when performing `init`?

This will help me in debugging further. Thank you :)

**Comment by namnx228** [comment]
Hi @ykakarap, sorry for my late answer.
After digging into the source code of clusterctl, here is my finding about how it uses the override directory in case of upgrade. To upgrade a specific provider:
- First, it checks if there is any information regarding this provider in `clusterctl.yaml` (from providers field as in comment above). If provider field is not provided, it checks the information from Github to get a list of versions that have been released.
- After that, it checks if the latest version is available in the override directory. If yes, it uses the override one and get the metadata.yaml from here. Otherwise, it fetches the metadata.yaml file from that latest release of that provider on Github.

To sum up, if we put a test version of provider (this version is not in the list of released versions), and we don't specify the use of this version in `clusterctl.yaml`, `clusterctl upgrade` will ignore the local override version and fetch the latest release from Github. 
This behavior is not really a bug of recent `clusterctl` version because the code looks like that for a long time (2 years!!!), but I suggest that we should update the doc to make it clear that provider field in `clusterctl.yaml` is needed to use override directory in upgrade workflow.

**Comment by fabriziopandini** [comment]
@namnx228 thanks for reporting the result of your investigation!

/remove-kind bug
/kind documentation

/retitle clarify how clusterctl overrides works
The point to stress in the doc is the fact that overrides only provide file replacements; instead, provider version resolution is based only on the actual repository structure

**Comment by fabriziopandini** [comment]
/good-first-issue

We should document the fact that overrides only provide file replacements; instead, provider version resolution is based only on the actual repository structure with a not at the end of https://cluster-api.sigs.k8s.io/clusterctl/configuration.html#overrides-layer
